src/concurrency/thread.rs

```diff
@@ -38,70 +38,62 @@ pub enum TlsAllocAction {
     Leak,
 }
 
-/// Trait for callbacks that are executed when a thread gets unblocked.
-pub trait UnblockCallback<'tcx>: VisitProvenance {
-    /// Will be invoked when the thread was unblocked the "regular" way,
-    /// i.e. whatever event it was blocking on has happened.
-    fn unblock(self: Box<Self>, ecx: &mut InterpCx<'tcx, MiriMachine<'tcx>>) -> InterpResult<'tcx>;
-
-    /// Will be invoked when the timeout ellapsed without the event the
-    /// thread was blocking on having occurred.
-    fn timeout(self: Box<Self>, _ecx: &mut InterpCx<'tcx, MiriMachine<'tcx>>)
-    -> InterpResult<'tcx>;
+/// The completion state of an asynchronous machine operation in Miri's concurrency system.
+/// Used by callbacks to determine how a blocked thread should proceed after being woken.
+#[derive(Debug)]
+pub enum MachineCallbackState {
+    /// The operation completed successfully and the thread can proceed with its normal execution.
+    Ready,
+    /// The operation did not complete within its specified duration and should handle the timeout case.
+    TimedOut,
+}
+
+/// Generic callback trait for machine operations.
+pub trait MachineCallback<'tcx>: VisitProvenance {
+    /// Called when the operation completes (successfully or with timeout).
+    fn call(
+        self: Box<Self>,
+        ecx: &mut InterpCx<'tcx, MiriMachine<'tcx>>,
+        result: MachineCallbackState,
```

Comment:
The argument type should not be hard-coded.
```suggestion
        arg: T,
```